smolyay/surrogate.py

```diff
@@ -0,0 +1,272 @@
+        self._coefficients = None
+
+    @property
+    def domain(self):
+        """numpy.ndarray: domain of the real function to be approximated."""
+        self._domain = numpy.array(self._domain)
+        return self._domain
+
+    @domain.setter
+    def domain(self, value):
+        self._domain = numpy.array(value, ndmin=2)
+        self._reset_grid()
+
+    @property
+    def grid_generator(self):
+        """:class:`IndexGridGenerator`: Generator used for making the grids."""
+        return self._grid_generator
+
+    @grid_generator.setter
+    def grid_generator(self, value):
+        if not isinstance(value, IndexGridGenerator):
+            raise TypeError('Grids must be IndexGridGenerator')
+        self._grid_generator = value
+        self._reset_grid()
+
+    @property
+    def grid(self):
+        """:class:`IndexGrid` data structure."""
+        if self._grid is None:
+            self._coefficients = None
+            self._grid = self.grid_generator(self.domain.shape[0])
+        return self._grid
+
+    def _reset_grid(self):
+        """Reset the points and basis functions if the generator is changed."""
+        self._grid = None
+
+    def _make_basis_matrix(self):
+        """Generate basis matrix.
+
+        Parameters
+        ----------
+        points: list
+            Grid points.
+        basis_functions: list
+            Basis function of grid points.
+
+        Returns
+        -------
+        numpy.ndarray
+            Basis matrix.
+        """
+        points, basis_functions = self.grid.points, self.grid.basis_functions
+        basis_matrix = numpy.zeros((len(points),
+                                    len(points)))
+        for i, point in enumerate(points):
+            for j, basis in enumerate(basis_functions):
+                basis_matrix[i, j] = numpy.prod(
+                    [f(x) for x, f in zip(point, basis)])
+        return basis_matrix
+
+    def train(self, real_function, linear_solver='lu'):
+        """Fit surrogate's components (basis functions) to the function.
+
+        Parameters
+        ----------
+        real_function: callabe
+            Function to be approximated.
+        linear_slover: string
+            Method of solving linear equations:
+                'lu': lower upper decomposition.
+                'inv': solve for coefficients through x = A^{-1}B.
+                'lstsq': least-square solution.
+
+        Returns
+        -------
+        list
+            Coefficients of the surrogate.
+        """
+        # transform grid points
+        transformed_grids = numpy.zeros((
+            numpy.array(self.grid.points).shape[0], self._domain.shape[0]))
+        for dimension in range(self._domain.shape[0]):
+            transformed_grids[:, dimension] = (
+                numpy.polynomial.polyutils.mapdomain(
+                    numpy.array(self.grid.points)[:, dimension],
+                    (-1, 1), self.domain[dimension]))
```

Comment:
This is clever that there is a NumPy tool for doing this, but it is a little limiting and unfortunate that it only works one dimension at a time. The equations from their documentation (https://numpy.org/doc/stable/reference/generated/numpy.polynomial.polyutils.mapdomain.html) are easy to generalize to an `(N,dim)` array though:

```
def _mapdomain(self, x, old, new):
    old = numpy.array(old, copy=False, ndmin=2)
    new = numpy.array(new, copy=False, ndmin=2)
    x = numpy.array(x, copy=False, ndmin=2)

    # error checking
    if old.shape != new.shape:
        raise TypeError('Old and new domain must have the same shape')
    if old.shape[1] != 2:
        raise TypeError('Domain should be a dim x 2 array')
    if old.shape[0] != x.shape[1]:
        raise TypeError('Points must be an N x dim array matching domain')

    new_x = new[:,0] + (new[:,1]-new[:,0])*((x-old[:,0])/(old[:,1]-old[:,0]))
    if old.shape[0] == 1:
        new_x = new_x[0]
    return new_x
```
This can be run in either direction
```
real_points = self._mapdomain(scaled_points, np.array([[-1, 1]]*self.dimension), self.domain)
scaled_points = self._mapdomain(real_points, self.domain, np.array([[-1, 1]]*self.dimension))
```